File: CSharp/src/FAI.Router/Training/JudgeTrainer.cs
using AI.ML.NeuralNetworks.V2;
using AI.ML.NeuralNetworks.V2.Losses;
using AI.ML.NeuralNetworks.V2.Nn;
using AI.ML.NeuralNetworks.V2.Optim;
using FAI.Router.JudgeLogic;

namespace FAI.Router.Training;

/// <summary>
/// Обучение судьи: матрица трансформации подгоняется так, чтобы оценка судьи повторяла
/// оценку человека. Ошибкой служит квадрат расхождения, а обучение идет спуском по градиенту.
/// </summary>
public class JudgeTrainer
{
    private readonly Judge _judge;
    private readonly float _learningRate;

    /// <summary>
    /// Обучение судьи по оценкам человека
    /// </summary>
    /// <param name="judge">Судья, чья матрица обучается</param>
    /// <param name="learningRate">Скорость обучения</param>
    public JudgeTrainer(Judge judge, float learningRate = 0.01f)
    {
        _judge = judge;
        _learningRate = learningRate;
    }

    /// <summary>
    /// Один шаг обучения. Возвращает ошибку до шага, по ее убыванию видно, что судья учится.
    /// </summary>
    /// <param name="requested">Заказанная спецификация (ТЗ)</param>
    /// <param name="actual">Фактическая спецификация ответа</param>
    /// <param name="humanScore">Оценка человека: единица означает «нравится», ноль означает «нет»</param>
    public double Train(Specifications requested, Specifications actual, double humanScore)
    {
        int dim = Settings.FeaturesSpecDim;

        // Тензор строится на шаг от НЫНЕШНЕЙ матрицы судьи, а не копируется один раз в
        // конструкторе. Копия переставала быть матрицей судьи после первой же загрузки весов:
        // Load подменяет объект матрицы, и первый шаг обучения перезаписывал загруженное
        // единичной матрицей с одним шагом. Спуск без импульса состояния между шагами не
        // держит, поэтому терять здесь нечего.
        Parameter transformer = new(TensorBridge.ToTensor(_judge.TransformerW));
        Optimizer optimizer = new SGD([transformer], lr: _learningRate);

        Tensor request = TensorBridge.ToColumn(requested.FeaturesSpecificationVector);
        Tensor fact = TensorBridge.ToRow(actual.FeaturesSpecificationVector);
        Tensor target = Tensor.From([(float)humanScore], new Shape(1));

        optimizer.ZeroGrad();

        Tensor transformed = transformer.Tensor.MatMul(request).Reshape(1, dim);
        Tensor score = EmbeddingLosses.CosineSimilarity(fact, transformed);
        Tensor loss = RegressionLosses.MSE(score, target);

        loss.Backward();
        optimizer.Step();

        // Судья считает по Matrix, а не по тензору: без обратной записи обучение
        // никак не отразилось бы на его оценках
        TensorBridge.WriteBack(transformer.Tensor, _judge.TransformerW);

        return TensorBridge.Scalar(loss);
    }
}

File: Python/fai_router/training/judge_trainer.py
from __future__ import annotations

import numpy as np

from fai_router.judge import Judge
from fai_router.specifications import Specifications


class JudgeTrainer:
    """Обучение судьи: матрица трансформации подгоняется так, чтобы оценка судьи повторяла
    оценку человека. Ошибкой служит квадрат расхождения, обучение идет спуском по градиенту.

    В версии на C# градиент считал автоград из AI.NeuralNetworks. Здесь он выведен аналитически:
    для косинуса между фактом f и преобразованным заказом u = W r производная по W равна
    внешнему произведению производной по u на r."""

    def __init__(self, judge: Judge, learning_rate: float = 0.01):
        self._judge = judge
        self._lr = learning_rate

    def train(self, requested: Specifications, actual: Specifications, human_score: float) -> float:
        """Один шаг обучения. Возвращает ошибку до шага: по ее убыванию видно, что судья учится."""
        eps = 1e-8
        r = requested.feature_vector()
        f = actual.feature_vector()
        w = self._judge.transformer_w

        u = w @ r
        norm_f = np.sqrt(f @ f + eps * eps)
        norm_u = np.sqrt(u @ u + eps * eps)
        score = (f @ u) / (norm_f * norm_u)
        loss = (score - human_score) ** 2

        d_score = 2 * (score - human_score)
        d_u = f / (norm_f * norm_u) - (f @ u) * u / (norm_f * norm_u ** 3)
        gradient = np.outer(d_score * d_u, r)

        self._judge.transformer_w = w - self._lr * gradient
        return float(loss)
